src/test/java/org/broadinstitute/hellbender/utils/downsampling/PositionalDownsamplerUnitTest.java

```diff
@@ -0,0 +1,207 @@
+                // Mix of mapped, pure unmapped, and mapped with assigned position
+                { Arrays.asList(createStackOfMappedReads(10, "1", 1), createStackOfUnmappedReadsWithPosition(10, "1", 1), createStackOfMappedReads(5, "2", 1), createStackOfUnmappedReadsWithPosition(5, "2", 1), createStackOfUnmappedReads(100)), 5, Arrays.asList(5, 5, 100), true },
+                // Unmapped only
+                { Arrays.asList(createStackOfUnmappedReads(10)), 1, Arrays.asList(10), true },
+                // No reads
+                { Collections.emptyList(), 1, Collections.emptyList(), false }
+        };
+    }
+
+    @Test(dataProvider = "PositionalDownsamplerTestData")
+    public void testPositionalDownsampler( final List<List<GATKRead>> reads, final int targetCoverage, final List<Integer> expectedStackSizes, final boolean lastStackIsPureUnmapped ) {
+        final List<GATKRead> allReads = new ArrayList<>();
+        for ( final List<GATKRead> readStack : reads ) {
+            allReads.addAll(readStack);
+        }
+
+        final ReadsDownsampler downsampler = new PositionalDownsampler(targetCoverage, header);
+
+        Utils.resetRandomGenerator();
+        downsampler.submit(allReads);
+
+        final boolean expectFinalizedItemsAfterSubmission = expectedStackSizes.size() > 1 || (expectedStackSizes.size() == 1 && lastStackIsPureUnmapped);
+        Assert.assertEquals(downsampler.hasFinalizedItems(), expectFinalizedItemsAfterSubmission);
+        if ( expectFinalizedItemsAfterSubmission ) {
+            Assert.assertNotNull(downsampler.peekFinalized());
+        }
+        else {
+            Assert.assertNull(downsampler.peekFinalized());
+        }
+
+        final boolean expectPendingItemsAfterSubmission = ! expectedStackSizes.isEmpty() && ! lastStackIsPureUnmapped;
+        Assert.assertEquals(downsampler.hasPendingItems(), expectPendingItemsAfterSubmission);
+        if ( expectPendingItemsAfterSubmission ) {
+            Assert.assertNotNull(downsampler.peekPending());
+        }
+        else {
+            Assert.assertNull(downsampler.peekPending());
+        }
+
+        downsampler.signalEndOfInput();
+
+        final boolean expectFinalizedItemsAfterEndOfInput = ! expectedStackSizes.isEmpty();
+        Assert.assertEquals(downsampler.hasFinalizedItems(), expectFinalizedItemsAfterEndOfInput);
+        if ( expectFinalizedItemsAfterEndOfInput ) {
+            Assert.assertNotNull(downsampler.peekFinalized());
+        }
+        else {
+            Assert.assertNull(downsampler.peekFinalized());
+        }
+
+        Assert.assertFalse(downsampler.hasPendingItems());
+        Assert.assertNull(downsampler.peekPending());
+
+        List<GATKRead> downsampledReads = downsampler.consumeFinalizedItems();
+        Assert.assertFalse(downsampler.hasFinalizedItems() || downsampler.hasPendingItems());
```

Comment:
this line `assertFalse(downsampler.hasFinalizedItems() || downsampler.hasPendingItems());` would be simpler to read as two lines: `assertFalse(downsampler.hasFinalizedItems()) ` and `assertFalse( downsampler.hasPendingItems())`